src/nodes.coffee

```diff
@@ -1030,18 +1097,69 @@ exports.Arr = class Arr extends Base
 #### Class
 
 # The CoffeeScript class definition.
-# Initialize a **Class** with its name, an optional superclass, and a
-# list of prototype property assignments.
+# Initialize a **Class** with its name, an optional superclass, and a body.
 exports.Class = class Class extends Base
+  children: [ 'variable', 'parent', 'body' ]
+
+  defaultClassVariableName: '_Class'
+
   constructor: (@variable, @parent, @body = new Block) ->
-    @boundFuncs = []
-    @body.classBody = yes
+    @properties = []
 
-  children: ['variable', 'parent', 'body']
+  compileNode: (o) ->
+    if jumpNode = @body.jumps()
+      jumpNode.error 'Class bodies cannot contain pure statements'
+    if argumentsNode = @body.contains isLiteralArguments
+      argumentsNode.error "Class bodies shouldn't reference arguments"
 
-  defaultClassVariableName: '_Class'
+    name = @determineName()
+
+    [ directives, properties ] = @walkBody name
+    @setContext name
+
+    # TODO Once ES super is being used throughout, the check for `@parent`, and the assignment
+    # below, can be removed.
+    emptyBody = not @parent and not @externalCtor and directives.length is 0 and
+      (@body.isEmpty() or @body.expressions.length <= properties.length)
+
+    if emptyBody
+      klass = new ClassDeclaration name, @parent, new Block properties
+    else
+      ident   = new IdentifierLiteral name
+      params  = []
+      args    = []
+      wrapper = new Code params, @body
+      klass   = new Call wrapper, args
+
+      @body.spaced = true
 
-  # Figure out the appropriate name for the constructor function of this class.
+      o.classScope = wrapper.makeScope o.scope
+
+      declaration = new ClassDeclaration name
+      declaration.body.push property.hoist() for property in properties
+
+      if @externalCtor
+        externalCtor = new IdentifierLiteral o.classScope.freeVariable 'ctor', reserve: no
+        declaration.externalCtor = externalCtor
+        @externalCtor.variable.base = externalCtor
+
+      if @parent
+        parent = new IdentifierLiteral o.classScope.freeVariable 'superClass', reserve: no
```

Comment:
Why are we passing in a reference to the parent class, rather than just extending it? In other words, why this:
```js
Horse = (function(superClass) {
  class Horse extends superClass {};
  Horse.__super__ = superClass.prototype;
  return Horse;
})(Animal);
```
instead of this:
```js
class Horse extends Animal {};
```
Is it because we need `__super__` until we refactor `super`?